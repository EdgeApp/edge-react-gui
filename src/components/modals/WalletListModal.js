// @flow

import * as React from 'react'
import { type AirshipBridge } from 'react-native-airship'

import { SPECIAL_CURRENCY_INFO } from '../../constants/WalletAndCurrencyConstants.js'
import { useHandler } from '../../hooks/useHandler.js'
import s from '../../locales/strings.js'
import { useMemo, useState } from '../../types/reactHooks.js'
import { useSelector } from '../../types/reactRedux.js'
import { type EdgeTokenId } from '../../types/types.js'
import { makeCurrencyCodeTable } from '../../util/utils.js'
import { showError } from '../services/AirshipInstance.js'
import { ModalCloseArrow, ModalTitle } from '../themed/ModalParts.js'
import { OutlinedTextInput } from '../themed/OutlinedTextInput.js'
import { ThemedModal } from '../themed/ThemedModal.js'
import { WalletList } from '../themed/WalletList.js'

export type WalletListResult = {
  walletId?: string,
  currencyCode?: string,
  tokenId?: string
}

type Props = {|
  bridge: AirshipBridge<WalletListResult>,

  // Filtering:
  allowedAssets?: EdgeTokenId[],
  excludeAssets?: EdgeTokenId[],
  excludeWalletIds?: string[],
  filterActivation?: boolean,
  allowKeysOnlyMode?: boolean,

  // Visuals:
  headerTitle: string,
  showCreateWallet?: boolean,

  // Deprecated. Use `allowedAssets` and `excludeAssets` instead.
  // Valid formats include "ETH", "REP", or "ETH-REP",
  // and an empty list is the same as an undefined list:
  allowedCurrencyCodes?: string[],
  excludeCurrencyCodes?: string[]
|}

const KeysOnlyModeTokenIds: EdgeTokenId[] = Object.keys(SPECIAL_CURRENCY_INFO)
  .filter(pluginId => SPECIAL_CURRENCY_INFO[pluginId].keysOnlyMode ?? false)
  .map(pluginId => ({
    pluginId
  }))

export function WalletListModal(props: Props) {
  const {
    bridge,

    // Filtering:
    allowedAssets,
    excludeAssets,
    excludeWalletIds,
    filterActivation,
    allowKeysOnlyMode = false,

    // Visuals:
    headerTitle,
    showCreateWallet,

    // Deprecated:
    allowedCurrencyCodes,
    excludeCurrencyCodes
  } = props

  const account = useSelector(state => state.core.account)
  const [searching, setSearching] = useState(false)
  const [searchText, setSearchText] = useState('')

  // Upgrade deprecated props:
  const [legacyAllowedAssets, legacyExcludeAssets] = useMemo(() => {
    if (allowedCurrencyCodes == null && excludeCurrencyCodes == null) return []

    const lookup = makeCurrencyCodeTable(account.currencyConfig)
    const allowedAssets = upgradeCurrencyCodes(lookup, allowedCurrencyCodes)
    const excludeAssets = upgradeCurrencyCodes(lookup, excludeCurrencyCodes)

    return [allowedAssets, excludeAssets]
  }, [account, allowedCurrencyCodes, excludeCurrencyCodes])

  const handleCancel = useHandler(() => {
    bridge.resolve({})
  })
  const handlePress = useHandler((walletId: string, currencyCode: string) => {
    if (walletId === '') {
      handleCancel()
      showError(s.strings.network_alert_title)
    } else bridge.resolve({ walletId, currencyCode })
  })
  const handleSearchClear = useHandler(() => {
    setSearchText('')
    setSearching(false)
  })
  const handleSearchUnfocus = useHandler(() => setSearching(searchText.length > 0))
  const handleSearchFocus = useHandler(() => setSearching(true))

  // Prevent plugins that are "watch only" from being used unless it's explicitly allowed
  const walletListExcludeAssets = useMemo(() => {
    const result = excludeAssets ?? legacyExcludeAssets
    return allowKeysOnlyMode ? result : KeysOnlyModeTokenIds.concat(result ?? [])
  }, [allowKeysOnlyMode, excludeAssets, legacyExcludeAssets])

  return (
    <ThemedModal bridge={bridge} onCancel={handleCancel}>
      <ModalTitle center>{headerTitle}</ModalTitle>
      <OutlinedTextInput
        returnKeyType="search"
        label={s.strings.search_wallets}
        onChangeText={setSearchText}
        onFocus={handleSearchFocus}
        onBlur={handleSearchUnfocus}
        onClear={handleSearchClear}
        value={searchText}
        marginRem={[0.5, 0.75, 1.25]}
        searchIcon
      />
      <WalletList
        allowedAssets={allowedAssets ?? legacyAllowedAssets}
        excludeAssets={walletListExcludeAssets}
        excludeWalletIds={excludeWalletIds}
        filterActivation={filterActivation}
        marginRem={listMargin}
        searching={searching}
        searchText={searchText}
        showCreateWallet={showCreateWallet}
        onPress={handlePress}
      />
      <ModalCloseArrow onPress={handleCancel} />
    </ThemedModal>
  )
}

const listMargin = [0, -1]

/**
 * Precisely identify the assets named by a currency-code array.
 * Accepts plain currency codes, such as "ETH" or "REP",
 * but also scoped currency codes like "ETH-REP".
 *
 * The goal is to delete this once the wallet stops using this legacy format
 * internally.
 */
export function upgradeCurrencyCodes(lookup: (currencyCode: string) => EdgeTokenId[], currencyCodes?: string[]): EdgeTokenId[] | void {
  if (currencyCodes == null || currencyCodes.length === 0) return

  const out: EdgeTokenId[] = []
  for (const currencyCode of currencyCodes) {
    const [parentCode, tokenCode] = currencyCode.split('-')

    if (tokenCode == null) {
      // It's a plain code, like "REP", so add all matches:
      out.push(...lookup(parentCode))
    } else {
      // It's a scoped code, like "ETH-REP", so filter using the parent:
      const parent = lookup(parentCode).find(match => match.tokenId == null)
      if (parent == null) continue
      out.push(...lookup(tokenCode).filter(match => match.pluginId === parent.pluginId))
    }
  }
  return out
}
