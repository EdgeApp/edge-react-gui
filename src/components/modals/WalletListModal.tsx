import { FlashList } from '@shopify/flash-list'
import { EdgeAccount } from 'edge-core-js'
import * as React from 'react'
import { TouchableOpacity, View } from 'react-native'
import { AirshipBridge } from 'react-native-airship'
import { sprintf } from 'sprintf-js'

import { SPECIAL_CURRENCY_INFO } from '../../constants/WalletAndCurrencyConstants'
import { PaymentMethodsMap } from '../../controllers/action-queue/WyreClient'
import { useAsyncValue } from '../../hooks/useAsyncValue'
import { useHandler } from '../../hooks/useHandler'
import s from '../../locales/strings'
import { config } from '../../theme/appConfig'
import { useSelector } from '../../types/reactRedux'
import { NavigationBase } from '../../types/routerTypes'
import { BooleanMap, EdgeTokenId } from '../../types/types'
import { getCurrencyCode } from '../../util/CurrencyInfoHelpers'
import { fixSides, mapSides, sidesToMargin } from '../../util/sides'
import { makeCurrencyCodeTable } from '../../util/utils'
import { CustomAsset } from '../data/row/CurrencyRow'
import { PaymentMethodRow } from '../data/row/PaymentMethodRow'
import { Airship, showError } from '../services/AirshipInstance'
import { useTheme } from '../services/ThemeContext'
import { EdgeText } from '../themed/EdgeText'
import { MainButton } from '../themed/MainButton'
import { ModalFooter, ModalTitle } from '../themed/ModalParts'
import { OutlinedTextInput } from '../themed/OutlinedTextInput'
import { ThemedModal } from '../themed/ThemedModal'
import { WalletList } from '../themed/WalletList'
import { WalletListCurrencyRow } from '../themed/WalletListCurrencyRow'
import { ButtonsModal } from './ButtonsModal'

export interface WalletListResult {
  currencyCode?: string
  tokenId?: string
  walletId?: string

  // Wyre buy/sell
  isBankSignupRequest?: boolean
  fiatAccountId?: string

  // Custom asset selection
  customAsset?: CustomAsset
}

interface Props {
  bridge: AirshipBridge<WalletListResult>
  navigation: NavigationBase

  // Filtering:
  allowedAssets?: EdgeTokenId[]
  allowedWalletIds?: string[]
  allowKeysOnlyMode?: boolean
  customAssets?: CustomAsset[]
  excludeAssets?: EdgeTokenId[]
  excludeWalletIds?: string[]
  filterActivation?: boolean

  // Visuals:
  createWalletId?: string
  headerTitle: string
  showBankOptions?: boolean
  showCreateWallet?: boolean

  // Deprecated. Use `allowedAssets` and `excludeAssets` instead.
  // Valid formats include "ETH", "REP", or "ETH-REP",
  // and an empty list is the same as an undefined list:
  allowedCurrencyCodes?: string[]
  excludeCurrencyCodes?: string[]
}

const KeysOnlyModeTokenIds: EdgeTokenId[] = Object.keys(SPECIAL_CURRENCY_INFO)
  .filter(pluginId => SPECIAL_CURRENCY_INFO[pluginId].keysOnlyMode ?? false)
  .map(pluginId => ({
    pluginId
  }))

export function WalletListModal(props: Props) {
  const {
    bridge,
    navigation,

    // Filtering:
    allowedAssets,
    allowKeysOnlyMode = false,
    allowedWalletIds,
    customAssets,
    excludeAssets,
    excludeWalletIds,
    filterActivation,

    // Visuals:
    createWalletId,
    headerTitle,
    showBankOptions = false,
    showCreateWallet,

    // Deprecated:
    allowedCurrencyCodes,
    excludeCurrencyCodes
  } = props

  // #region Constants

  const showCustomAssets = customAssets != null && customAssets.length > 0

  const account = useSelector(state => state.core.account)
  const theme = useTheme()

  // #endregion Constants

  // #region State

  const [searching, setSearching] = React.useState(false)
  const [searchText, setSearchText] = React.useState('')

  const [bankAccountsMap] = useAsyncValue(async (): Promise<PaymentMethodsMap | null> => {
    // TODO: Re-enable once new fiat ramp partner is re-integrated
    return null
  }, [account])

  // #endregion State

  // #region Init

  // Upgrade deprecated props
  const [legacyAllowedAssets, legacyExcludeAssets] = React.useMemo(() => {
    if (allowedCurrencyCodes == null && excludeCurrencyCodes == null) return []

    const lookup = makeCurrencyCodeTable(account.currencyConfig)
    const allowedAssets = upgradeCurrencyCodes(lookup, allowedCurrencyCodes)
    const excludeAssets = upgradeCurrencyCodes(lookup, excludeCurrencyCodes)

    return [allowedAssets, excludeAssets]
  }, [account, allowedCurrencyCodes, excludeCurrencyCodes])

  // Prevent plugins that are "watch only" from being used unless it's explicitly allowed
  const walletListExcludeAssets = React.useMemo(() => {
    const result = excludeAssets ?? legacyExcludeAssets
    return allowKeysOnlyMode ? result : KeysOnlyModeTokenIds.concat(result ?? [])
  }, [allowKeysOnlyMode, excludeAssets, legacyExcludeAssets])

  // #endregion Init

  // #region Handlers

  const handleCancel = useHandler(() => {
    bridge.resolve({})
  })
  const handlePaymentMethodPress = useHandler((fiatAccountId: string) => () => {
    bridge.resolve({ fiatAccountId })
  })
  const handleWalletListPress = useHandler((walletId: string, currencyCode: string, _tokenId?: string, customAsset?: CustomAsset) => {
    if (walletId === '') {
      handleCancel()
      showError(s.strings.network_alert_title)
    } else bridge.resolve({ walletId, currencyCode, customAsset })
  })
  const handleSearchClear = useHandler(() => {
    setSearchText('')
    setSearching(false)
  })
  const handleSearchUnfocus = useHandler(() => setSearching(searchText.length > 0))
  const handleSearchFocus = useHandler(() => setSearching(true))

  // Pull up the signup workflow on the calling scene if the user does not yet have a linked bank account
  const handleShowBankPlugin = useHandler(async () => {
    const result = await Airship.show<'continue' | undefined>(bridge => (
      <ButtonsModal
        bridge={bridge}
        title={s.strings.deposit_to_bank}
        message={sprintf(s.strings.wallet_list_modal_confirm_s_bank_withdrawal, config.appName)}
        buttons={{
          continue: { label: s.strings.legacy_address_modal_continue }
        }}
      />
    ))
    if (result === 'continue') await bridge.resolve({ isBankSignupRequest: true })
  })

  // #endregion Handlers

  // #region Renderers

  const renderBankSignupButton = () => (
    <MainButton label={s.strings.deposit_to_bank} type="secondary" onPress={handleShowBankPlugin} marginRem={[0, 0.75, 1.5, 0.75]} />
  )

  const renderPaymentMethod = useHandler(item => {
    return (
      <TouchableOpacity onPress={handlePaymentMethodPress(item.item.id)}>
        <PaymentMethodRow paymentMethod={item.item} pluginId="wyre" key={item.item.id} />
      </TouchableOpacity>
    )
  })

  const renderCustomAsset = useHandler(item => {
    return <WalletListCurrencyRow wallet={item.item.wallet} tokenId={item.tokenId} customAsset={item.item} onPress={handleWalletListPress} />
  })

  const renderBankSection = () => {
    if (bankAccountsMap == null) return null
    if (!showBankOptions) return null
    if (Object.keys(bankAccountsMap).length === 0) return renderBankSignupButton()
    return (
      <View>
        <FlashList
          estimatedItemSize={theme.rem(4.25)}
          data={Object.values(bankAccountsMap)}
          keyboardShouldPersistTaps="handled"
          renderItem={renderPaymentMethod}
          keyExtractor={item => item.id}
          style={sidesToMargin(mapSides(fixSides([-1, -1, 1, -0.5], 0), theme.rem))}
        />
      </View>
    )
  }

  const renderCustomAssetSection = () =>
    showCustomAssets ? (
      <View>
        <FlashList
          estimatedItemSize={theme.rem(4.25)}
          data={customAssets}
          keyboardShouldPersistTaps="handled"
          renderItem={renderCustomAsset}
          keyExtractor={item => item.referenceTokenId}
          style={sidesToMargin(mapSides(fixSides([-0.5, -1, 1, -1], 0), theme.rem))}
        />
      </View>
    ) : null

  // #endregion Renderers

  return (
    <ThemedModal bridge={bridge} onCancel={handleCancel}>
      <ModalTitle center>{headerTitle}</ModalTitle>
      {renderBankSection()}
      {renderCustomAssetSection()}
      {showBankOptions || showCustomAssets ? <EdgeText>{s.strings.your_wallets}</EdgeText> : null}
      <OutlinedTextInput
        returnKeyType="search"
        label={s.strings.search_wallets}
        onChangeText={setSearchText}
        onFocus={handleSearchFocus}
        onBlur={handleSearchUnfocus}
        onClear={handleSearchClear}
        value={searchText}
        marginRem={[0.5, 0, 1.25, 0]}
        searchIcon
      />
      <WalletList
        allowedAssets={allowedAssets ?? legacyAllowedAssets}
        allowedWalletIds={allowedWalletIds}
        excludeAssets={walletListExcludeAssets}
        excludeWalletIds={excludeWalletIds}
        filterActivation={filterActivation}
        marginRem={listMargin}
        searching={searching}
        searchText={searchText}
        showCreateWallet={showCreateWallet}
        createWalletId={createWalletId}
        onPress={handleWalletListPress}
        navigation={navigation}
      />
      <ModalFooter onPress={handleCancel} fadeOut />
    </ThemedModal>
  )
}

const listMargin = [0, -1]

/**
 * Precisely identify the assets named by a currency-code array.
 * Accepts plain currency codes, such as "ETH" or "REP",
 * but also scoped currency codes like "ETH-REP".
 *
 * The goal is to delete this once the wallet stops using this legacy format
 * internally.
 */
export function upgradeCurrencyCodes(lookup: (currencyCode: string) => EdgeTokenId[], currencyCodes?: string[]): EdgeTokenId[] | undefined {
  if (currencyCodes == null || currencyCodes.length === 0) return

  const out: EdgeTokenId[] = []
  for (const currencyCode of currencyCodes) {
    const [parentCode, tokenCode] = currencyCode.split('-')

    if (tokenCode == null) {
      // It's a plain code, like "REP", so add all matches:
      out.push(...lookup(parentCode))
    } else {
      // It's a scoped code, like "ETH-REP", so filter using the parent:
      const parent = lookup(parentCode).find(match => match.tokenId == null)
      if (parent == null) continue
      out.push(...lookup(tokenCode).filter(match => match.pluginId === parent.pluginId))
    }
  }
  return out
}

// Given a list of assets, shows a modal for a user to pick a wallet for that asset.
// If only one wallet exists for that asset, auto pick that wallet
export const pickWallet = async ({
  account,
  allowedWalletIds,
  assets,
  headerTitle = s.strings.select_wallet,
  navigation,
  showCreateWallet
}: {
  account: EdgeAccount
  allowedWalletIds?: string[]
  assets?: EdgeTokenId[]
  headerTitle?: string
  navigation: NavigationBase
  showCreateWallet?: boolean
}): Promise<WalletListResult | undefined> => {
  const { currencyWallets } = account

  const walletIdMap: BooleanMap = {}

  // Check if user owns any wallets that
  const matchingAssets = (assets ?? []).filter(asset => {
    const matchingWalletIds: string[] = Object.keys(currencyWallets).filter(key => {
      const { pluginId, tokenId } = asset
      const currencyWallet = currencyWallets[key]
      const pluginIdMatch = currencyWallet.currencyInfo.pluginId === pluginId

      // No wallet with matching pluginId, fail this asset
      if (!pluginIdMatch) return false
      if (tokenId == null) {
        walletIdMap[key] = true
        return true
      }
      // See if this wallet has a matching token enabled
      const tokenIdMatch = currencyWallet.enabledTokenIds.find(tid => tokenId)
      if (tokenIdMatch != null) {
        const cc = getCurrencyCode(currencyWallet, tokenIdMatch)
        walletIdMap[`${key}:${cc}`] = true
        return true
      }
      return false
    })
    return matchingWalletIds.length !== 0
  })

  if (assets != null && matchingAssets.length === 0) return

  if (assets != null && matchingAssets.length === 1 && Object.keys(walletIdMap).length === 1) {
    // Only one matching wallet and asset. Auto pick the wallet
    const [walletId, currencyCode] = Object.keys(walletIdMap)[0].split(':')
    return { walletId, currencyCode }
  } else {
    const walletListResult = await Airship.show<WalletListResult>(bridge => (
      <WalletListModal
        bridge={bridge}
        navigation={navigation}
        headerTitle={headerTitle}
        allowedWalletIds={allowedWalletIds}
        allowedAssets={assets}
        showCreateWallet={showCreateWallet}
      />
    ))
    return walletListResult
  }
}
