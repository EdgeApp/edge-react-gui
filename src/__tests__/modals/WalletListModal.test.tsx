import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import TestRenderer from 'react-test-renderer'

import { upgradeCurrencyCodes, WalletListModal } from '../../components/modals/WalletListModal'
import { EdgeTokenId } from '../../types/types'
import { fakeAirshipBridge } from '../../util/fake/fakeAirshipBridge'
import { fakeNavigation } from '../../util/fake/fakeNavigation'
import { FakeProviders } from '../../util/fake/FakeProviders'

describe('WalletListModal', () => {
  it('should render with loading props', () => {
    const renderer = TestRenderer.create(
      <FakeProviders>
        <WalletListModal bridge={fakeAirshipBridge} navigation={fakeNavigation} headerTitle="Wallet List" />
      </FakeProviders>
    )

    expect(renderer.toJSON()).toMatchSnapshot()
  })

  it("Should upgrade currency codes to token ID's", () => {
    const data: { [code: string]: EdgeTokenId[] } = {
      ETH: [{ pluginId: 'ethereum' }],
      BNB: [{ pluginId: 'binance' }, { pluginId: 'ethereum', tokenId: '1234abcd' }]
    }
    function lookup(currencyCode: string): EdgeTokenId[] {
      return data[currencyCode.toUpperCase()] ?? []
    }

    // Check ambiguous currency codes:
    expect(upgradeCurrencyCodes(lookup, ['ETH', 'BNB'])).toEqual([
      { pluginId: 'ethereum' },
      { pluginId: 'binance' },
      { pluginId: 'ethereum', tokenId: '1234abcd' }
    ])

    // Check scoped currency codes:
    expect(upgradeCurrencyCodes(lookup, ['ETH', 'ETH-BNB'])).toEqual([{ pluginId: 'ethereum' }, { pluginId: 'ethereum', tokenId: '1234abcd' }])

    // Check missing currency codes:
    expect(upgradeCurrencyCodes(lookup, ['ETH', 'LOL'])).toEqual([{ pluginId: 'ethereum' }])
  })
})
