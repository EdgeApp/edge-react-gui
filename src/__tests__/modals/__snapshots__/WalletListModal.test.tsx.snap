// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletListModal should render with loading props 1`] = `
Array [
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "transparent",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <BlurView
      blurAmount={10}
      blurType="light"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>,
  <View
    collapsable={false}
    style={
      Object {
        "alignSelf": "flex-end",
        "backgroundColor": "#121d25",
        "borderColor": "rgba(255, 255, 255, 0)",
        "borderTopLeftRadius": 22,
        "borderTopRightRadius": 22,
        "borderWidth": 0,
        "flexDirection": undefined,
        "flexShrink": 1,
        "justifyContent": undefined,
        "marginBottom": -64,
        "marginLeft": 0,
        "marginRight": 0,
        "marginTop": 0,
        "maxHeight": undefined,
        "overflow": "visible",
        "paddingBottom": 86,
        "paddingLeft": 22,
        "paddingRight": 22,
        "paddingTop": 22,
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 10,
        "transform": Array [
          Object {
            "translateY": 1334,
          },
        ],
        "width": 512,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "margin": 11,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Quicksand-Medium",
              "fontSize": 27,
              "marginVertical": 11,
            },
            Object {
              "textAlign": "center",
            },
            Object {
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
            },
          ]
        }
      >
        Wallet List
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "rgba(255, 255, 255, 0)",
            "borderRadius": 11,
            "justifyContent": "center",
            "minHeight": 67,
            "paddingHorizontal": 22,
          },
          Object {
            "flexGrow": 0,
            "marginBottom": 28,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 11,
            "paddingLeft": 65,
            "paddingRight": 65,
            "paddingVertical": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "borderTopWidth": 1,
              "bottom": 0,
              "left": 22,
              "position": "absolute",
              "right": 22,
            },
            Object {
              "borderColor": undefined,
              "right": 11,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "borderBottomLeftRadius": 11,
              "borderBottomWidth": 1,
              "borderLeftWidth": 1,
              "borderRightWidth": 0,
              "borderTopLeftRadius": 11,
              "borderTopWidth": 1,
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "top": 0,
              "width": 22,
            },
            Object {
              "borderColor": undefined,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "borderBottomRightRadius": 11,
              "borderBottomWidth": 1,
              "borderLeftWidth": 0,
              "borderRightWidth": 1,
              "borderTopRightRadius": 11,
              "borderTopWidth": 1,
              "bottom": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 22,
            },
            Object {
              "borderColor": undefined,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "borderTopWidth": 1,
              "left": 22,
              "position": "absolute",
              "right": 22,
              "top": 0,
            },
            Object {
              "borderColor": undefined,
              "left": 22,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "height": 67,
              "justifyContent": "center",
              "left": 0,
              "paddingHorizontal": 22,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "paddingLeft": 65,
              "paddingRight": 65,
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          onLayout={[Function]}
          style={
            Array [
              Object {
                "alignSelf": "flex-start",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "padding": 0,
              },
              Object {
                "color": undefined,
                "transform": Array [
                  Object {
                    "translateX": -0,
                  },
                  Object {
                    "translateY": -0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
              },
            ]
          }
        >
          Search Wallets
        </Text>
      </View>
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "bottom": -28,
              "color": "#E85466",
              "fontFamily": "Quicksand-Regular",
              "fontSize": 17,
              "left": 28,
              "position": "absolute",
            },
            Object {
              "opacity": 0,
            },
          ]
        }
      />
      <Text
        numberOfLines={1}
        onLayout={[Function]}
        style={
          Array [
            Object {
              "bottom": -10,
              "fontFamily": "Quicksand-Regular",
              "fontSize": 17,
              "position": "absolute",
              "right": 28,
            },
            Object {
              "color": undefined,
              "opacity": 0,
              "transform": Array [
                Object {
                  "translateX": 6,
                },
                Object {
                  "scale": 0,
                },
              ],
            },
          ]
        }
      >
        
      </Text>
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 12,
            },
            Object {
              "color": "rgba(255, 255, 255, .75)",
              "fontSize": 22,
              "left": 0,
              "padding": 22,
              "position": "absolute",
              "top": 0,
            },
            Object {
              "fontFamily": "anticon",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        î™°
      </Text>
      <TextInput
        autoFocus={false}
        editable={true}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        returnKeyType="search"
        secureTextEntry={false}
        selectionColor="#FFFFFF"
        style={
          Array [
            Object {
              "alignSelf": "stretch",
              "color": "#FFFFFF",
              "fontFamily": "Quicksand-Regular",
              "fontSize": 22,
              "padding": 0,
            },
            Object {
              "flexGrow": 0,
            },
          ]
        }
        textAlignVertical="top"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "overflow": "hidden",
        }
      }
    >
      <RCTScrollView
        applyWindowCorrection={[Function]}
        canChangeSize={true}
        contentContainerStyle={
          Object {
            "backgroundColor": undefined,
            "minHeight": 1,
            "minWidth": 1,
            "paddingLeft": 0,
            "paddingRight": 0,
          }
        }
        contentHeight={0}
        contentWidth={0}
        data={Array []}
        dataProvider={
          DataProvider {
            "_data": Array [],
            "_firstIndexToProcess": 0,
            "_hasStableIds": false,
            "_requiresDataChangeHandling": false,
            "_size": 0,
            "getStableId": [Function],
            "rowHasChanged": [Function],
          }
        }
        disableRecycling={false}
        estimatedItemSize={96}
        extendedState={Object {}}
        externalScrollView={[Function]}
        finalRenderAheadOffset={250}
        forceNonDeterministicRendering={true}
        horizontal={false}
        initialOffset={0}
        initialRenderIndex={0}
        isHorizontal={false}
        keyboardShouldPersistTaps="handled"
        layoutProvider={
          GridLayoutProviderWithProps {
            "_acceptableRelayoutDelta": 1,
            "_getHeightOrWidth": [Function],
            "_getLayoutTypeForIndex": [Function],
            "_getSpan": [Function],
            "_hasExpired": false,
            "_maxSpan": 1,
            "_setLayoutForType": [Function],
            "_tempDim": Object {
              "height": 0,
              "width": 0,
            },
            "averageWindow": AverageWindow {
              "currentAverage": 96,
              "currentCount": 1,
              "inputValues": Array [
                96,
              ],
              "nextIndex": 1,
            },
            "defaultEstimatedItemSize": 100,
            "layoutObject": Object {
              "size": undefined,
              "span": undefined,
            },
            "props": Object {
              "data": Array [],
              "estimatedItemSize": 96,
              "keyboardShouldPersistTaps": "handled",
              "numColumns": 1,
              "renderItem": [Function],
              "style": Object {
                "marginBottom": 0,
                "marginLeft": -22,
                "marginRight": -22,
                "marginTop": 0,
              },
            },
            "renderWindowInsets": Object {
              "height": 0,
              "width": 0,
            },
            "shouldRefreshWithAnchoring": true,
          }
        }
        maxRenderAhead={750}
        numColumns={1}
        onEndReached={[Function]}
        onEndReachedThreshold={0}
        onEndReachedThresholdRelative={0}
        onItemLayout={[Function]}
        onLayout={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onSizeChanged={[Function]}
        onVisibleIndicesChanged={[Function]}
        removeClippedSubviews={false}
        renderAheadOffset={0}
        renderAheadStep={250}
        renderContentContainer={[Function]}
        renderItem={[Function]}
        renderItemContainer={[Function]}
        rowRenderer={[Function]}
        scrollEventThrottle={16}
        scrollThrottle={16}
        scrollViewProps={
          Object {
            "contentContainerStyle": Object {
              "backgroundColor": undefined,
              "minHeight": 1,
              "minWidth": 1,
              "paddingLeft": 0,
              "paddingRight": 0,
            },
            "onLayout": [Function],
            "onScrollBeginDrag": [Function],
            "refreshControl": undefined,
            "style": Object {
              "minHeight": 1,
              "minWidth": 1,
            },
          }
        }
        style={
          Object {
            "minHeight": 1,
            "minWidth": 1,
          }
        }
        suppressBoundedSizeException={true}
        windowCorrectionConfig={
          Object {
            "applyToInitialOffset": true,
            "applyToItemScroll": true,
            "value": Object {
              "endCorrection": 0,
              "startCorrection": 0,
              "windowShift": -0,
            },
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "flexDirection": "column",
              }
            }
          >
            <View
              style={
                Object {
                  "paddingLeft": 0,
                  "paddingTop": 0,
                }
              }
            />
            <View
              style={
                Array [
                  undefined,
                  undefined,
                ]
              }
            />
            <AutoLayoutView
              enableInstrumentation={false}
              horizontal={false}
              onBlankAreaEvent={[Function]}
              onLayout={[Function]}
              renderAheadOffset={0}
              scrollOffset={0}
              style={
                Object {
                  "height": 0,
                  "width": 0,
                }
              }
              windowSize={0}
            />
            <CellContainer
              index={-1}
              style={
                Array [
                  undefined,
                  undefined,
                ]
              }
            />
            <View
              style={
                Object {
                  "paddingBottom": 0,
                  "paddingRight": 0,
                }
              }
            />
          </View>
        </View>
      </RCTScrollView>
    </View>
    <View
      style={
        Object {
          "marginBottom": -22,
        }
      }
    >
      <View
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityHint="Close Modal"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
              "padding": 22,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#00f1a2",
                  "fontSize": 28,
                },
                undefined,
                Object {
                  "fontFamily": "anticon",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            î˜³
          </Text>
        </View>
      </View>
      <BVLinearGradient
        colors={
          Array [
            1187109,
            856825125,
            3205635365,
            4279377189,
          ]
        }
        endPoint={
          Object {
            "x": 0,
            "y": 1,
          }
        }
        locations={
          Array [
            0,
            0.2,
            0.75,
            1,
          ]
        }
        pointerEvents="none"
        startPoint={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        style={
          Object {
            "bottom": 0,
            "height": 67,
            "position": "absolute",
            "width": "100%",
          }
        }
      />
    </View>
  </View>,
]
`;
